app/scripts/lib/rpc-method-middleware/handlers/wallet-createSession/handler.ts

```diff
@@ -0,0 +1,248 @@
+      validateAndNormalizeScopes(requiredScopes || {}, optionalScopes || {});
+
+    const requiredScopesWithSupportedMethodsAndNotifications =
+      getSupportedScopeObjects(normalizedRequiredScopes);
+    const optionalScopesWithSupportedMethodsAndNotifications =
+      getSupportedScopeObjects(normalizedOptionalScopes);
+
+    const networkClientExistsForChainId = (chainId: Hex) => {
+      try {
+        hooks.findNetworkClientIdByChainId(chainId);
+        return true;
+      } catch (err) {
+        return false;
+      }
+    };
+
+    const { supportedScopes: supportedRequiredScopes } = bucketScopes(
+      requiredScopesWithSupportedMethodsAndNotifications,
+      {
+        isChainIdSupported: networkClientExistsForChainId,
+        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+      },
+    );
+
+    const { supportedScopes: supportedOptionalScopes } = bucketScopes(
+      optionalScopesWithSupportedMethodsAndNotifications,
+      {
+        isChainIdSupported: networkClientExistsForChainId,
+        isChainIdSupportable: () => false, // intended for future usage with eip3085 scopedProperties
+      },
+    );
+
+    // Fetch EVM accounts from native wallet keyring
+    // These addresses are lowercased already
+    const existingEvmAddresses = hooks
+      .listAccounts()
+      .map((account) => account.address);
+    const supportedEthAccounts = getEthAccounts({
+      requiredScopes: supportedRequiredScopes,
+      optionalScopes: supportedOptionalScopes,
+    })
+      .map((address) => address.toLowerCase() as Hex)
+      .filter((address) => existingEvmAddresses.includes(address));
+
+    const requestedCaip25CaveatValue = {
+      requiredScopes: getInternalScopesObject(supportedRequiredScopes),
+      optionalScopes: getInternalScopesObject(supportedOptionalScopes),
+      isMultichainOrigin: true,
+    };
+
+    const requestedCaip25CaveatValueWithSupportedEthAccounts = setEthAccounts(
+      requestedCaip25CaveatValue,
+      supportedEthAccounts,
+    );
+
+    const { permissions: approvedPermissions } =
+      await hooks.requestPermissionApprovalForOrigin({
+        [Caip25EndowmentPermissionName]: {
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: requestedCaip25CaveatValueWithSupportedEthAccounts,
+            },
+          ],
+        },
+      });
+
+    const approvedCaip25Permission =
+      approvedPermissions[Caip25EndowmentPermissionName];
+    const approvedCaip25CaveatValue = approvedCaip25Permission?.caveats?.find(
+      (caveat) => caveat.type === Caip25CaveatType,
+    )?.value as Caip25CaveatValue;
+    if (!approvedCaip25CaveatValue) {
+      throw rpcErrors.internal();
+    }
+
+    const sessionScopes = getSessionScopes(approvedCaip25CaveatValue);
+
+    hooks.grantPermissions({
+      subject: {
+        origin,
+      },
+      approvedPermissions,
+    });
```

Comment:
We should modify this to use the `PermissionController.requestPermissions` directly rather than ApprovalController -> `PermissionController.grantPermissions`

Same idea as the refactor we did here: https://github.com/MetaMask/metamask-extension/pull/30042/files#diff-6fbff2cfe97ac01b77296ef2122c7e0a5b3ff6a84b584b4d1a87482f35eea3d6

Reply:
Done here: https://github.com/MetaMask/metamask-extension/pull/27782/commits/b710fbb98468f92f5280bb9ced79aaaad45e15e2